controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentStatus.java

```diff
@@ -107,6 +110,51 @@ public JobList jobs() {
         return allJobs;
     }
 
+    /** Whether any jobs both dependent on the dependency, and a dependency for the dependent, are failing. */
+    private boolean hasFailures(StepStatus dependency, StepStatus dependent) {
+        Set<StepStatus> dependents = new HashSet<>();
+        fillDependents(dependency, new HashSet<>(), dependents, dependent);
+        Set<JobId> criticalJobs = dependents.stream().flatMap(step -> step.job().stream()).collect(Collectors.toSet());
+
+        return ! allJobs.matching(job -> criticalJobs.contains(job.id()))
+                        .failing()
+                        .not().withStatus(RunStatus.outOfCapacity)
+                        .isEmpty();
+    }
+
+    private boolean fillDependents(StepStatus dependency, Set<StepStatus> visited, Set<StepStatus> dependents, StepStatus current) {
+        if (visited.contains(current))
+            return dependents.contains(current);
+
+        if (dependency == current) {
+            dependents.add(current);
+            visited.add(current);
+            return true;
+        }
+
+        for (StepStatus dep : current.dependencies)
+            if (fillDependents(dependency, visited, dependents, dep))
+                dependents.add(current);
+        visited.add(current);
+
+        return dependents.contains(current);
+    }
+
+    /** Whether any jobs of the given instance, or in an instance dependent on it, are failing with other errors than lack of capacity in a test zone. */
+    public boolean hasFailures(InstanceName instance) {
+        Set<StepStatus> dependents = new HashSet<>();
+        for (StepStatus instanceStep : instanceSteps().values())
```

Comment:
Hmm, this should use all steps as starting points, not just all instance steps.

Reply:
Urgh, this also doesn't catch implicit tests ...

Reply:
Perhaps the logic should instead be to check where there are any failures with an older revision than what we're considering deploying ...

Reply:
... and just check all jobs. Tests are tricky, because they can be run under any instance—there could even be tests in an earlier instance that are failing because of a platform-revision combination in a later instance that just wasn't present in any of the earlier ones, so the "dependent instances" idea break down. :'( 

But considering revision ordering should always be correct for this use case ... I think!